flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/e2e/KafkaSourceE2ECase.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.kafka.source.e2e;
+
+import org.apache.flink.connector.kafka.source.e2e.resources.KafkaContainerizedExternalSystem;
+import org.apache.flink.connector.kafka.source.e2e.resources.KafkaMultipleTopicExternalContext;
+import org.apache.flink.connector.kafka.source.e2e.resources.KafkaSingleTopicExternalContext;
+import org.apache.flink.connectors.test.common.environment.FlinkContainersTestEnvironment;
+import org.apache.flink.connectors.test.common.environment.MiniClusterTestEnvironment;
+import org.apache.flink.connectors.test.common.junit.annotations.WithExternalContextFactory;
+import org.apache.flink.connectors.test.common.junit.annotations.WithExternalSystem;
+import org.apache.flink.connectors.test.common.junit.annotations.WithTestEnvironment;
+import org.apache.flink.connectors.test.common.testsuites.TestSuiteBase;
+import org.apache.flink.connectors.test.common.utils.ConnectorJarUtils;
+
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Tag;
+
+/**
+ * End-to-end test for {@link org.apache.flink.connector.kafka.source.KafkaSource} using testing
+ * framework, based on JUnit 5.
+ *
+ * <p>This class uses {@link Nested} classes to test functionality of KafkaSource under different
+ * Flink environments (MiniCluster and Containers). Each nested class extends {@link TestSuiteBase}
+ * for reusing test cases already defined by testing framework.
+ */
+@Disabled
+@Tag("TestingFramework")
+@DisplayName("Kafka Source E2E Test")
+public class KafkaSourceE2ECase {
+
+    @Nested
+    @DisplayName("On MiniCluster")
+    class OnMiniCluster extends TestSuiteBase<String> {
```

Comment:
I'd probably rather go for a parameterized method that supplies a collection of TestEnvironments. Then we could even use a different set of `TestEnvironments` depending on where it's executed. It's also significantly fewer LOC.

Reply:
I split IT and E2E cases into two class. IT case (on MiniCluster) stays in ```flink-connector-kafka``` module and E2E case (on Docker) goes to ```flink-end-to-end-tests-common-kafka```. This also solves the Flink image building issues above since E2E case can reused the existing ```FlinkContainer``` that can build the Flink image based on flink-dist on the fly.

Reply:
I think this is a good solution for now but I'd like to revisit it later; it would be lead to a module explosion if we add a new e2e module for all connectors under test. But for Kafka, this works well.